### ISSUE TITLE
Bus stops and shelters

### ISSUE THREAD

**Issue by Robot8A** [gravitystorm/openstreetmap-carto#2431]
Hi, I've noticed that when you map the shelter next to a bus stop, on the render the shelter icon always overrides the bus stop icon. I consider that the bus stop icon is significantly more important that the shelter one for the general public.
An example can be found here: http://www.openstreetmap.org/?mlat=41.66628&mlon=-0.88670#map=18/41.66629/-0.88670

You could do that when a shelter is placed next to a bus stop (or when a shelter that has the tag shelter_type=public_transport) the bus icon is displayed before the shelter icon.
Thanks in advance.

**Comment by HolgerJeromin** [comment]
Another tagging options is the attribute [shelter](http://wiki.openstreetmap.org/wiki/Key:shelter) on the bus_stop. But a detailed mapping should be possible, yes.

**Comment by manfredbrandl** [comment]
@HolgerJeromin here the shelter is mapped as an area and the bus_stop is mapped as a point. Do you have an idea how to combine these two features?

**Comment by Robot8A** [comment]
@manfredbrandl There is a stop_area relation: https://www.openstreetmap.org/relation/5552716

**Comment by Robot8A** [comment]
Here you have another example: https://www.openstreetmap.org/relation/5519522

**Comment by Robot8A** [comment]
Key:public_transport in the wiki: https://wiki.openstreetmap.org/wiki/Tag:public_transport%3Dstop_area#Members

**Comment by kocio-pl** [comment]
Is anybody ready to add the code which shows bus shelters on z19+? I'm willing to help if somebody likes to try.

**Comment by Tomasz-W** [comment]
related to #964

**Comment by DerDings** [comment]
Just an idea, I don't know if this is possible:

Currently blue is used for public transport features, but grey for shelters. Since the shelters mentioned here are also public transport features, I think we should make them blue if they are tagged `public_transport=yes`, otherwise keep the current colour.
In many places, like [this example](https://www.openstreetmap.org/#map=18/51.43655/7.37856), ![grafik](https://user-images.githubusercontent.com/42044179/105210515-6e87cb00-5b4b-11eb-8d37-1734f37839cc.png) this will look very similar to bus stop squares.

As an optional, second step, the shelters rendering importance may be reduced. Since bus stop shelters and other shelters are different things now, this probably can be done easily for the bus stop shelters only, without modifying the other shelter types. Thus, its less likely that this change has unwanted results elsewhere.

**Comment by masterofnoroad** [comment]
I suppose someone thought about it before, but maybe it would be better to just make special version of bus stop icon with shelter? Something like OSMAnd is doing. It could also render in similar way public transport shelters.

When I'm on the road cycling and sometimes I need to find shelter from the rain fast, I prefer to have shelter icon rendered but bus stop icon is more important generally :D

![142004926_264999038356534_7080675383208415647_n](https://user-images.githubusercontent.com/72669347/105773255-eb5be000-5f63-11eb-8782-39739151c787.jpg)

**Comment by DerDings** [comment]
> I suppose someone thought about it before, but maybe it would be better to just make special version of bus stop icon with shelter? Something like OSMAnd is doing. It could also render in similar way public transport shelters.
> 
> When I'm on the road cycling and sometimes I need to find shelter from the rain fast, I prefer to have shelter icon rendered but bus stop icon is more important generally :D
> 
> ![142004926_264999038356534_7080675383208415647_n](https://user-images.githubusercontent.com/72669347/105773255-eb5be000-5f63-11eb-8782-39739151c787.jpg)
> 
> 

Are you talking about one object tagged as `highway=bus_stop shelter =yes`? 
#3124 is about this case.

If there are two separate objects, one `highway=bus_stop` and one `amenity=shelter shelter_type=public transport` it would truly be useful to show that as you suggested, but I think it is not efficiently possible combining those two features during rendering.

**Comment by masterofnoroad** [comment]
Yes, indeed I was thinking highway=bus_stop shelter =yes. Its a pity that it was not introduced.

If combining two separate objects is hard to implement, it would be maybe better to come back to #3124 solution and then prioritize bus stop with shelter icon over amenity=shelter? 
When Im reading the discussion I understand it died on lack of good readable and similar to existing bus stop icon. So maybe it would be better to just work little bit more on the design and not drop the idea entirely? It doesnt need to be different icon, it just need to be existing bus stop icon + small roof on top of it as I showed in OSMand.

**Comment by jajajaneeneenee** [comment]
Some notes to the first message in this issue of 2016:

> Hi, I've noticed that when you map the shelter next to a bus stop, on the render the shelter icon always overrides the bus stop icon. I consider that the bus stop icon is significantly more important that the shelter one for the general public.
> An example can be found here: http://www.openstreetmap.org/?mlat=41.66628&mlon=-0.88670#map=18/41.66629/-0.88670

@Robot8A: The bus_stop icon is shown in this example in zoom level 19. Has it been moved further away from the shelter since your posting in 2016? Or did you want to see it also in zoom level 18 (or lower)? Or did anything change in the style since 2016? 

I have an example from my city where the bus_stop icon is not even shown in zoom level 19: [https://www.openstreetmap.org/node/273533270](https://www.openstreetmap.org/node/273533270).
(The distance from the node to the shelter is only 0.6 m – its position is perhaps not perfectly correct, because normally it's a distance of at least 1 or 2 m – but perhaps it's a good example ...).

> You could do that when a shelter is placed next to a bus stop (or when a shelter that has the tag shelter_type=public_transport) the bus icon is displayed before the shelter icon.
> Thanks in advance.

I am not very familiar with the restrictions of the rendering in openstreet map carto and its style, and I am not 100% sure what you meant by _"the bus icon is displayed **before** the shelter icon"_ and I wanted to ask a maybe stupid question (perhaps this is also what you meant with "before"?):

Why can the bus_stop icon not be rendered **over** the shelter area (and its icon), and if it's very near to the shelter, also with an overlapping? This also happens with other icons which are very near, e.g. `natural=tree` and `tourism=information/information=guidepost` oder `natural=tree` and `historic=memorial` and surely others. Is this a problem for the styling?

It would look like this – and wouldn't that be ok? (And easy to implement in the style? – I don't know ... sorry.) And much better than no bus stop icon at all?

![bus_stop_and_shelter_72dpi](https://user-images.githubusercontent.com/48966999/114802001-2c160900-9d9d-11eb-993f-28fb81fc5180.png)
(Some variants depending on the position of the bus_stop node; and this shelter has a size of 3 x 1.5 m).

And  if shelters with `shelter_type=public_transport` would be painted in a **blue (cyan) colour** – the same as the bus icon – it would look like this:

![bus_stop_and_shelter_blue_icon_blue_72dpi](https://user-images.githubusercontent.com/48966999/114804322-6bdeef80-9da1-11eb-8bfe-f174041fbd5e.png)

Or like this (if the shelter icon keeps its grey colour):

![bus_stop_and_shelter_blue_icon_grey_72dpi](https://user-images.githubusercontent.com/48966999/114804379-7e592900-9da1-11eb-946a-0b256d7ab57b.png)

But the visually best variant is probably the first ... And also with regard to the importance of the two objects – because the bus_stop icon is in the foreground visually.

**Comment by imagico** [comment]
> Why can the bus_stop icon not be rendered over the shelter area (and its icon), and if it's very near to the shelter, also with an overlapping?

Because Mapnik collision handling is too primitive for that.  You can either place a symbol adhering to collision detection (i.e. not place it if it would collide) or place it ignoring collisions.  But you can't place it allowing collisions with some other symbols but not with others.  I quickly looked into that in context of drinking water sources - see [here](http://blog.imagico.de/the-way-of-the-water/).  I think the only way you could do what you indicate would be to generate aggregate svg symbols outside of mapnik.

Independent of these technical issues your idea would only work if you implement some sort of displacement to avoid cases where the bus stop and the shelter symbol would be placed directly on top of each other which would not be readable.

**Comment by DerDings** [comment]
> ![bus_stop_and_shelter_72dpi](https://user-images.githubusercontent.com/48966999/114802001-2c160900-9d9d-11eb-993f-28fb81fc5180.png)
> (Some variants depending on the position of the bus_stop node; and this shelter has a size of 3 x 1.5 m).
> 
> And if shelters with `shelter_type=public_transport` would be painted in a **blue (cyan) colour** – the same as the bus icon – it would look like this:
> 
> ![bus_stop_and_shelter_blue_icon_blue_72dpi](https://user-images.githubusercontent.com/48966999/114804322-6bdeef80-9da1-11eb-8bfe-f174041fbd5e.png)
> 
> Or like this (if the shelter icon keeps its grey colour):
> 
> ![bus_stop_and_shelter_blue_icon_grey_72dpi](https://user-images.githubusercontent.com/48966999/114804379-7e592900-9da1-11eb-946a-0b256d7ab57b.png)
> 

The idea was to _only_ make the icon blue.

The area of the shelter currently isn't rendered at all. The brown area in your example is the rendering of `building=*`. It is only because someone added `building=*` to that shelter. In my opinion, the building's colour shouldn't be altered.

Blue shelter icon + blue bus stop icon would look acceptable in my opinon, if the current minimal collison distance is not altered.

**Comment by bgo-eiu** [comment]
I agree with this

**Comment by BertMule** [comment]
Please change this.
I ran into exactly the same problem, and wanted to report it.
This issue is already 7 years old.

Shelters are hiding the bus-stops, of which only the name is visible (if any).
If one is looking unsuspectedly for a bus-stop, one won't find it.

Her is a cluster of 3 stops plus shelters, none is visible (slightly depending on zoom level).
https://www.openstreetmap.org/#map=19/51.44462/4.92706

**Comment by kaneap** [comment]
Would it be an acceptable solution to simply render bus stops with higher priority than other amenity points? I was able to achieve this fairly easily in my fork of the repo. Before:
![before bus](https://github.com/gravitystorm/openstreetmap-carto/assets/35740782/a49523ca-2914-4b6b-ae03-02181b0ffd9a)
After:
![after bus](https://github.com/gravitystorm/openstreetmap-carto/assets/35740782/047aa118-ee5e-4f45-be69-01ce9bb96bd6)

**Comment by imagico** [comment]
> Would it be an acceptable solution to simply render bus stops with higher priority than other amenity points? I was able to achieve this fairly easily in my fork of the repo.

Changes to amenity-points that make it - fully or partially - adhere to prioritization by starting zoom level would be seen favorably.  Specifically prioritizing bus stops over amenity=shelter (same starting zoom level) would be fine as well (but not really feasible without giving everything else a consistent order of priority as well).

Prioritizing bus stops over everything else, however, would not be seen favorably because it would cause more problems than it solves.

See  #3880, #4676.

**Comment by k-yle** [comment]
It would be nice to apply the same logic to ferry terminals as well as bus stops. 

[Example:](https://osm.org/#map=18/-36.84531/175.15022) the transit shelter obscures the ⚓ icon
![image](https://github.com/gravitystorm/openstreetmap-carto/assets/16009897/9223e2a9-89c3-4de9-b29a-592b8a97ffab)

**Comment by Elefant-aus-Wuppertal** [comment]
Regardless of this, I could also imagine that `amenity=shelter` with `shelter_type=public_transport` should only be rendered in the last two zoom levels.

**Comment by dch0ph** [comment]
Yes, I've been testing the above and it works nicely. A public transport shelter is something small that you take advantage of, rather than a shelter you head to from a distance. It does clutter the map to render these from Z15. 

Happy to submit a PR if general enthusiasm.

**Comment by imagico** [comment]
Please note this issue is about the relative prioritization of shelters and bus stops.  Changing the starting zoom level of `shelter_type=public_transport` is covered in #3483, see also #3124 and #4317.

**Comment by dch0ph** [comment]
Thanks for the links. Suggest consolidating discussion on #3483, since this seems the route with the most overall support.

**Comment by maro-21** [comment]
It's been almost 10 years since this bug was reported, and still nothing.

The stop names aren't showing up if there is a shelter. The bus stop name is more important than the path or the shelter. After adding bus shelters, the stop names disappear from the map. This kind of thing discourages people from mapping details. I reported this myself 5 years ago, and it still hasn't been fixed.

https://osm.org/#map=19/51.053279/17.056937
<img width="294" height="252" alt="Image" src="https://github.com/user-attachments/assets/0ff991e5-7a2b-4792-a35f-a88f2ccb4234" />

**Comment by imagico** [comment]
> It's been almost 10 years since this bug was reported, and still nothing.

It is really inexcusable that the OSM-Carto developers who get paid so generously for working on the style - oh, wait, that was a different project...

This issue is about prioritizing the symbols.  Labels are a different matter.  Labels are currently rendered separately from symbols (relates to #234) and bus stop labels specifically relates to #4851.  Prioritizing some labels over symbols would require moving the symbols to `amenity-low-priority`.  But we generally deliberately prioritize symbols over labels that belong to symbols so we provide feedback on a feature being mapped (the shelter) with priority over additional details on a feature that is shown anyway (the name label of a bus stop).  That - for some use cases - this is not beneficial is an accepted consequence.  We are not a dedicated public transport map after all.
